src/training/validator.h

```diff
@@ -37,7 +37,8 @@ class ValidatorBase : public TrainingObserver {
 public:
   ValidatorBase(bool lowerIsBetter) : lowerIsBetter_(lowerIsBetter), lastBest_{initScore()} {}
 
-  virtual float validate(const std::vector<Ptr<ExpressionGraph>>& graphs) = 0;
+  virtual float validate(const std::vector<Ptr<ExpressionGraph>>& graphs,
+                         const TrainingState& state) = 0;
```

Comment:
I think we can use `Ptr<TrainingState>` here, following the general convention?

Reply:
You mean like in ValidatorBase::actAfterLoaded()? Done.